python/IECore/Enum.py

```diff
@@ -77,10 +79,20 @@ def __hash__( self ) :
 
 			return hash( ( self.__class__, self.__value ) )
 
-		def __cmp__( self, other ) :
+		def __eq__( self, other ):
+			try:
+				return  self.__value == other.__value
+			except AttributeError:
```

Comment:
Out of interest, why do we catch AttributeError? Doesn't the constructor guarantee that `__value` is always there? Or is this about comparing against other types - do we need to do that? The little bit of reading I did on Python 3 comparisons implied that the new way is to throw for mismatched types.

Reply:
I did it to replace a type assert that was failing after I changed to "new style" classes. I am not entirely sure what's the best approach here until all Cortex classes have been ported to the new style. The attribute approach should work fine in all cases, but I am not completely satisfied with it myself, since you could technically cheat it.

Reply:
I presume you mean the old `assert( type( self ) is type( other ) )` that is now removed? My point was that the catching AttributeError is the opposite of what we did before. It looks like we used to throw if the types didn't match, but now we silently return False. The [tiny bit of reading](https://portingguide.readthedocs.io/en/latest/comparisons.html#unorderable-types) I did around Python3 comparisons was that the convention is to throw, so it seemed odd that we were moving in the opposite direction.

Reply:
After deriving `Enum` from `object`, the current assertion code is definitely failing. When trying to launch `gaffer browser` it crashes with the following trace (abridged):

```
File "/software/apps/gaffer/0.54.2.2/cent7.x86_64/cortex/10/gaffer/python/GafferUI/SplitContainer.py", line 71, in setOrientation
    v = QtCore.Qt.Horizontal if orientation == self.Orientation.Horizontal else QtCore.Qt.Vertical
  File "/home/timle/apps/cortex/10.0.0-a68dev/cent7.x86_64/base/python/2.7/gcc/4.8.3/IECore/Enum.py", line 84, in __eq__
    assert ( type( self ) is type( other ) )
```

Checking the types of the compared objects reveals they are both Enum classes, but since Enum is defined inside a function call, Python treats the new style classes `<class 'IECore.Enum.Enum'>` as two different classes and the assertion fails.

Reply:
I think I can fix it using a metaclass and make the Enum more elegant at the same time .. after the holidays :-)